feat(linter): add `oxc/no-async-endpoint-handlers`

DonIsaac opened the pull request:
Adds `no-async-endpoint-handlers` rules, which bans async functions used as endpoint handlers in Express applications. These do not get caught by Express' error handler, causing the server to crash with an unhandled process rejection error.

```js
app.use(async (req, res) => {
  const foo = await api.getFoo(req.query) // server panics if this function rejects
  return res.json(foo)
})
```

I could not find this rule implemented in any ESLint plugin, but this is a problem I see quite often and I'm tired of dealing with it. I've added it to `oxc` for now, but we should consider adding an `express` or `api` plugin in the future.

camc314 commented on crates/oxc_linter/src/utils/express.rs:
```diff
@@ -0,0 +1,126 @@
+        _ => return false,
+    };
+
+    // NOTE(@DonIsaac): should we check for destructuring patterns? I don't
+    // really ever see them used in handlers, and their existence could indicate
+    // this function is not a handler.
+    if params.rest.is_some() {
+        return false;
+    }
```
i wouldn't bother, i've never seen that used 🙂

DonIsaac replied:
Me neither, which is why its existence could indicate that the function is not a handler

camc314 replied:
yep -probably can remove the note?

DonIsaac replied:
Should we introduce this check?

camc314 replied:
i don't think so?

camc314 commented on crates/oxc_linter/src/rules/oxc/no_async_endpoint_handlers.rs:
```diff
@@ -0,0 +1,376 @@
+    use crate::tester::Tester;
+    use serde_json::json;
+
+    let pass = vec![
```
should we allow if there's a `try` `catch` in the stmt?

DonIsaac replied:
This could introduce a lot of difficulties with checking _what_ is being caught
```ts
async handle(req, res) {
  try {
   await couldThrow()
  } catch {}
  await couldAlsoThrow()
}
// or this example, which is never caught
async handle(req, res) {
  try {
    couldThrow.then(data => res.json(data))
  } catch {}
}
```
Making sure everything is covered by a try-catch sounds painful...

camc314 replied:
hmm i see. this seems reasonable to skip checking

DonIsaac commented:
* **#5364** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5364?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac committed:
feat(linter): add `oxc/no-async-endpoint-handlers` (#5364)

Adds `no-async-endpoint-handlers` rules, which bans async functions used as endpoint handlers in Express applications. These do not get caught by Express' error handler, causing the server to crash with an unhandled process rejection error.

```js
app.use(async (req, res) => {
  const foo = await api.getFoo(req.query) // server panics if this function rejects
  return res.json(foo)
})
```

I could not find this rule implemented in any ESLint plugin, but this is a problem I see quite often and I'm tired of dealing with it. I've added it to `oxc` for now, but we should consider adding an `express` or `api` plugin in the future.
